File: src/io.rs
// Copyright 2022 Datafuse Labs.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::io::Error;
use std::io::Result;

use bytes::Bytes;
use futures::AsyncRead;
use futures::AsyncWrite;
use futures::Sink;
use futures::Stream;

/// BytesRead represents a reader of bytes.
pub trait BytesRead: AsyncRead + Unpin + Send {}
impl<T> BytesRead for T where T: AsyncRead + Unpin + Send {}

/// BytesReader is a boxed dyn [`BytesRead`].
pub type BytesReader = Box<dyn BytesRead>;

/// BytesWrite represents a writer of bytes.
pub trait BytesWrite: AsyncWrite + Unpin + Send {}
impl<T> BytesWrite for T where T: AsyncWrite + Unpin + Send {}

/// BytesWriter is a boxed dyn [`BytesWrite`].
pub type BytesWriter = Box<dyn BytesWrite>;

/// BytesStream represents a stream of bytes.
///
/// This trait is used as alias to `Stream<Item = Result<Bytes>> + Unpin + Send`.
pub trait BytesStream: Stream<Item = Result<Bytes>> + Unpin + Send {}
impl<T> BytesStream for T where T: Stream<Item = Result<Bytes>> + Unpin + Send {}

/// BytesStreamer is a boxed dyn [`BytesStream`].
pub type BytesStreamer = Box<dyn BytesStream>;

/// BytesSink represents a sink of bytes.
///
/// THis trait is used as alias to `Sink<Bytes, Error = Error> + Unpin + Send`.
pub trait BytesSink: Sink<Bytes, Error = Error> + Unpin + Send {}
impl<T> BytesSink for T where T: Sink<Bytes, Error = Error> + Unpin + Send {}

/// BytesSinker is a boxed dyn [`BytesSink`]
pub type BytesSinker = Box<dyn BytesSink>;

File: bindings/ruby/lib/opendal_ruby/io.rb
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# frozen_string_literal: true

module OpenDAL
  class IO
    # Reads all lines from the stream into an array.
    # @raise [EOFError] when the end of the file is reached.
    # @return [Array<String>]
    def readlines
      results = []

      loop do
        results << readline
      rescue EOFError
        break
      end

      results
    end

    # Rewinds the stream to the beginning.
    def rewind
      seek(0, ::IO::SEEK_SET)
    end

    # Sets the file position to `new_position`.
    # @param new_position [Integer]
    def pos=(new_position)
      seek(new_position, ::IO::SEEK_SET)
    end

    alias_method :pos, :tell

    # Checks if the stream is at the end of the file.
    # @return [Boolean]
    def eof
      position = tell
      seek(0, ::IO::SEEK_END)
      tell == position
    end

    alias_method :eof?, :eof

    # Returns the total length of the stream.
    # @return [Integer]
    def length
      current_position = tell
      seek(0, ::IO::SEEK_END)
      tell.tap { self.pos = current_position }
    end

    alias_method :size, :length
  end
end
